File: biblemark/static/js/spa/model/markedVerse.js
import Verse from './verse.js';

export default class MarkedVerse {

    /**
     * @param {string} id
     * @param {Verse} verse
     */
    constructor(id, verse) {
        this.id = id;
        this.verse = verse;
    }

    /**
     * @param {string} id
     * @param {string} versionId
     * @param {string} bookId
     * @param {string} chapterId
     * @param {string} verseNumber
     * @returns {MarkedVerse}
     */
    static factory = (id, versionId, bookId, chapterId, verseNumber) =>
        new MarkedVerse(id, new Verse(versionId, bookId, chapterId, verseNumber));

    /**
     * @param {VersionedVerseId} versionedVerseId
     * @returns {MarkedVerse}
     */
    static fromVersionedVerseId = (versionedVerseId) =>
        new MarkedVerse(null, Verse.fromVersionedVerseId(versionedVerseId));
}

File: biblemark/model/marked_verse.py
from biblemark.exceptions.mark_http_exceptions import InvalidMarkedVerse
from biblemark.model.bible_verse import BibleVerse
from biblemark.model.bible_version import Version


class MarkedVerse:

    def __init__(self, verse: BibleVerse, visibility: bool = True, entity_id=None):
        self.entity_id = entity_id
        self.verse = verse

        if not isinstance(visibility, bool):
            raise InvalidMarkedVerse("No boolean indicator of visibility assigned to mark reference")
        self.visibility = visibility

    @staticmethod
    def factory(version: Version, book_id, chapter_id, verse_number, visibility=True, entity_id=None):
        return MarkedVerse(BibleVerse(version, book_id, chapter_id, verse_number), visibility, entity_id)

    def __str__(self) -> str:
        return f"MarkedVerse(id={self.entity_id}, verse={self.verse}, visibility={self.visibility})"

    def __repr__(self) -> str:
        return f"MarkedVerse(id={self.entity_id!r}, verse={self.verse!r}, visibility={self.visibility!r})"
